File: dev/src/samples/NetSamples/Samples/CogoPoints/PointGroupCommands.cs
using System;
using System.Collections.Generic;

using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.EditorInput;
using Autodesk.AutoCAD.Runtime;
using Autodesk.Civil.DatabaseServices;


[assembly: CommandClass(
    typeof(Autodesk.CivilizedDevelopment.PointGroupCommands))]

namespace Autodesk.CivilizedDevelopment
{
    public class PointGroupCommands : SimpleDrawingCommand
    {
        [CommandMethod("CDS_CreateCivilDevPointGroups")]
        public void CDS_CreateCivilDevPointGroups()
        {
            using (Transaction tr = startTransaction())
            {
                createPointGroup("Trees", "TREE*");
                createPointGroup("Wells", "WELL*");
                tr.Commit();
            }
        }

        [CommandMethod("CDS_ReversPointGroupDrawOrder")]
        public void CDS_ReversPointGroupDrawOrder()
        {
            ObjectIdCollection drawPriority = _pointGroups.DrawOrder;
            ObjectIdCollection reversed = new ObjectIdCollection();
            for (int i = drawPriority.Count - 1; i >= 0; i-- )
            {
                ObjectId current = drawPriority[i];
                reversed.Add(current);
            }
            _pointGroups.DrawOrder = reversed;
        }

        [CommandMethod("CDS_RenumberPointsForGroup")]
        public void CDS_RenumberPointsForGroup()
        {
            string pointGroupName = getPointGroupName();
            if (pointGroupName == String.Empty)
            {
                return;
            }

            int baseNumber = getNewBaseNumber();
            if (baseNumber == _kNoNumber)
            {
                return;
            }

            renumberPointsForGroup(pointGroupName, baseNumber);
        }

        [CommandMethod("CDS_CreatePointGroupsWithQueries")]
        public void CDS_CreatePointGroupsWithQueries()
        {
            using (Transaction tr = startTransaction())
            {
                createStandardPointGroup();
                createCustomPointGroup();
                tr.Commit();
            }
        }

        private void createStandardPointGroup()
        {
            StandardPointGroupQuery standard =
                    new StandardPointGroupQuery();
            standard.IncludeElevations = ">100.5";
            standard.IncludeFullDescriptions = "Contains*";
            standard.IncludeNames = "Tree Cedar*";
            standard.IncludeNumbers = "<1000";
            standard.IncludeRawDescriptions = "TREE*";
            standard.ExcludeElevations = ">300.0";
            standard.ExcludeFullDescriptions = "Contains POLE*";
            standard.ExcludeNames = "Tree maples*";
            standard.ExcludeNumbers = "<200";
            standard.ExcludeRawDescriptions = "POLE*";
            createPointGroup("Standard Group", standard);
        }

        private void createCustomPointGroup()
        {
            string queryString =
                    @"(PointNumber>=100 AND PointNumber<200) OR 
                        (FullDescription='Contains*' OR RawDescription='WE*')";
            CustomPointGroupQuery custom = new CustomPointGroupQuery();
            custom.QueryString = queryString;
            createPointGroup("Custom Group", custom);
        }

        private void createPointGroup(string name, PointGroupQuery query)
        {
            if (_pointGroups.Contains(name))
            {
                return;
            }
            ObjectId groupId = _pointGroups.Add(name);
            PointGroup group = groupId.GetObject(OpenMode.ForRead)
                as PointGroup;
            group.SetQuery(query);
        }

        private void createPointGroup(string name, 
            string includeRawDescription)
        {
            if (_pointGroups.Contains(name))
            {
                return;
            }
            ObjectId groupId = _pointGroups.Add(name);
            StandardPointGroupQuery query = new StandardPointGroupQuery();
            query.IncludeRawDescriptions = includeRawDescription;
            PointGroup group = groupId.GetObject(OpenMode.ForWrite) 
                as PointGroup;
            group.SetQuery(query);
            // NOTE:    Setting a description throws an exception. The
            //          issue is being researched by the API team.
            // group.Description = "Set a description";
        }

        private PointGroupCollection _pointGroups
        {
            get
            {
                return _civildoc.PointGroups;
            }
        }

        private string getPointGroupName()
        {
            PromptResult result = _editor.GetString(
                "\nEnter point group name: ");
            if (result.Status == PromptStatus.OK)
            {
                return result.StringResult;
            }
            return String.Empty;
        }

        private int getNewBaseNumber()
        {
            PromptIntegerResult result = _editor.GetInteger(
                "\nEnter new base number: ");
            if (result.Status == PromptStatus.OK)
            {
                return result.Value;
            }

            return _kNoNumber;
        }

        private void renumberPointsForGroup(string groupName, int baseNumber)
        {
            using (Transaction tr = startTransaction())
            {
                ObjectId pointGroupId = getPointGroupIdByName(groupName);
                PointGroup group = pointGroupId.GetObject(OpenMode.ForRead)
                    as PointGroup;
                renumberPoints(group, baseNumber);
                tr.Commit();
            }
        }

        private ObjectId getPointGroupIdByName(string groupName)
        {
            return _pointGroups[groupName];
        }

        private void renumberPoints(PointGroup group, int baseNumber)
        {
            uint[] pointNumbers = group.GetPointNumbers();
            int firstNumber = (int)pointNumbers[0];
            int factor = baseNumber - firstNumber;
            _civildoc.CogoPoints.SetPointNumber(
                ToEnumerableObjectId(group.GetPointNumbers()), factor);
            group.Update();
        }

        private IEnumerable<ObjectId> ToEnumerableObjectId(uint[] numbers)
        {
            foreach(uint number in numbers)
            {
                yield return _civildoc.CogoPoints
                    .GetPointByPointNumber(number);
            }
        }

        private const int _kNoNumber = -1;
    }
};

File: dev/src/samples/NetSamplesVB/Samples/CogoPoints/PointGroupCommands.vb
Imports System.Collections.Generic

Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.EditorInput
Imports Autodesk.AutoCAD.Runtime
Imports Autodesk.Civil.DatabaseServices

<Assembly: CommandClass(
GetType(Autodesk.CivilizedDevelopment.PointGroupCommands))> 

Namespace Autodesk.CivilizedDevelopment
  Public Class PointGroupCommands
    Inherits SimpleDrawingCommand
    <CommandMethod("CDS_CreateCivilDevPointGroups")> _
    Public Sub CDS_CreateCivilDevPointGroups()
      Using tr As Transaction = startTransaction()
        createPointGroup("Trees", "TREE*")
        createPointGroup("Wells", "WELL*")
        tr.Commit()
      End Using
    End Sub

    <CommandMethod("CDS_ReversPointGroupDrawOrder")> _
    Public Sub CDS_ReversPointGroupDrawOrder()
      Dim drawPriority As ObjectIdCollection = _pointGroups.DrawOrder
      Dim reversed As New ObjectIdCollection()
      For i As Integer = drawPriority.Count - 1 To 0 Step -1
        Dim current As ObjectId = drawPriority(i)
        reversed.Add(current)
      Next
      _pointGroups.DrawOrder = reversed
    End Sub

    <CommandMethod("CDS_RenumberPointsForGroup")> _
    Public Sub CDS_RenumberPointsForGroup()
      Dim pointGroupName As String = getPointGroupName()
      If pointGroupName = [String].Empty Then
        Return
      End If

      Dim baseNumber As Integer = getNewBaseNumber()
      If baseNumber = _kNoNumber Then
        Return
      End If

      renumberPointsForGroup(pointGroupName, baseNumber)
    End Sub

    <CommandMethod("CDS_CreatePointGroupsWithQueries")> _
    Public Sub CDS_CreatePointGroupsWithQueries()
      Using tr As Transaction = startTransaction()
        createStandardPointGroup()
        createCustomPointGroup()
        tr.Commit()
      End Using
    End Sub

    Private Sub createStandardPointGroup()
      Dim standard As New StandardPointGroupQuery()
      standard.IncludeElevations = ">100.5"
      standard.IncludeFullDescriptions = "Contains*"
      standard.IncludeNames = "Tree Cedar*"
      standard.IncludeNumbers = "<1000"
      standard.IncludeRawDescriptions = "TREE*"
      standard.ExcludeElevations = ">300.0"
      standard.ExcludeFullDescriptions = "Contains POLE*"
      standard.ExcludeNames = "Tree maples*"
      standard.ExcludeNumbers = "<200"
      standard.ExcludeRawDescriptions = "POLE*"
      createPointGroup("Standard Group", standard)
    End Sub

    Private Sub createCustomPointGroup()
      Dim queryString As String = "(PointNumber>=100 AND PointNumber<200) OR " _
        & vbCr & vbLf & "(FullDescription='Contains*' OR RawDescription='WE*')"
      Dim [custom] As New CustomPointGroupQuery()
      [custom].QueryString = queryString
      createPointGroup("Custom Group", [custom])
    End Sub

    Private Sub createPointGroup(name As String, query As PointGroupQuery)
      If _pointGroups.Contains(name) Then
        Return
      End If
      Dim groupId As ObjectId = _pointGroups.Add(name)
      Dim group As PointGroup = TryCast(groupId.GetObject(OpenMode.ForRead), 
        PointGroup)
      group.SetQuery(query)
    End Sub


    Private Sub createPointGroup(name As String,
                                 includeRawDescription As String)
      If _pointGroups.Contains(name) Then
        Return
      End If
      Dim groupId As ObjectId = _pointGroups.Add(name)
      Dim query As New StandardPointGroupQuery()
      query.IncludeRawDescriptions = includeRawDescription
      Dim group As PointGroup =
        TryCast(groupId.GetObject(OpenMode.ForRead), PointGroup)
      group.SetQuery(query)
      ' NOTE:     Setting a description throws an exception. The
      '           issue is being researched by the API team.
      ' group.Description = "Set a description";
    End Sub

    Private ReadOnly Property _pointGroups() As PointGroupCollection
      Get
        Return _civildoc.PointGroups
      End Get
    End Property

    Private Function getPointGroupName() As String
      Dim result As PromptResult =
        _editor.GetString(vbLf & "Enter point group name: ")
      If result.Status = PromptStatus.OK Then
        Return result.StringResult
      End If
      Return [String].Empty
    End Function

    Private Function getNewBaseNumber() As Integer
      Dim result As PromptIntegerResult =
        _editor.GetInteger(vbLf & "Enter new base number: ")
      If result.Status = PromptStatus.OK Then
        Return result.Value
      End If

      Return _kNoNumber
    End Function

    Private Sub renumberPointsForGroup(groupName As String,
                                       baseNumber As Integer)
      Using tr As Transaction = startTransaction()
        Dim pointGroupId As ObjectId = getPointGroupIdByName(groupName)
        Dim group As PointGroup =
          TryCast(pointGroupId.GetObject(OpenMode.ForRead), PointGroup)
        renumberPoints(group, baseNumber)
        tr.Commit()
      End Using
    End Sub

    Private Function getPointGroupIdByName(groupName As String) As ObjectId
      Return _pointGroups(groupName)
    End Function

    Private Sub renumberPoints(group As PointGroup, baseNumber As Integer)
      Dim pointNumbers As UInteger() = group.GetPointNumbers()
      Dim firstNumber As Integer = CInt(pointNumbers(0))
      Dim factor As Integer = baseNumber - firstNumber
      _civildoc.CogoPoints.SetPointNumber(
        ToEnumerableObjectId(group.GetPointNumbers()), factor)
      group.Update()
    End Sub

    Private Function ToEnumerableObjectId(numbers As UInteger()) _
        As IEnumerable(Of ObjectId)
      Dim points As List(Of ObjectId) = New List(Of ObjectId)
      For Each number As UInteger In numbers
        Dim id = _civildoc.CogoPoints.GetPointByPointNumber(number)
        points.Add(id)
      Next
      Return points
    End Function

    Private Const _kNoNumber As Integer = -1
  End Class
End Namespace
